### PR TITLE
Fix for 667

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

File: invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceWithMetadata.java [hunk 1]
```diff
@@ -69,11 +86,63 @@ private void build() {
...
         }
     }
 
+
+    public static  final Map<UUID, DateTime> computeChargedThroughDates(final Invoice invoice, final InternalCallContext context)  {
+        final Map<UUID, DateTime> chargedThroughDates;
+        if (invoice != null &&
+            invoice.getStatus() == InvoiceStatus.COMMITTED) /* See https://github.com/killbill/killbill/issues/1296 */ {
+            chargedThroughDates = new HashMap<>();
+                for (final InvoiceItem item : invoice.getInvoiceItems()) {
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceWithMetadata.java:121]: Maybe keep the original comment:

```
// Don't use invoice.getInvoiceItems(final Class<T> clazz) as some items can come from plugins
```

(it's an API we should probably remove at some point)

File: invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java [hunk 2]
```diff
@@ -651,17 +653,8 @@ private InvoiceWithFutureNotifications processAccountWithLockAndInputTargetDate(
                 // Commit invoice on disk
                 final ExistingInvoiceMetadata existingInvoiceMetadata = new ExistingInvoiceMetadata(existingInvoices);
                 commitInvoiceAndSetFutureNotifications(account, invoiceModelDao, billingEvents, trackingIds, futureAccountNotifications, existingInvoiceMetadata, internalCallContext);
+                setChargedThroughDatesNoExceptions(invoiceWithMetadata.getChargeThroughDates(), internalCallContext);
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java:656]: Should we really ignore the exception here?

File: beatrix/src/test/java/org/killbill/billing/beatrix/integration/usage/TestWithoutZeroUsageItems.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+public class TestWithoutZeroUsageItems extends TestIntegrationBase {
+
+
+    @Inject
+    protected InvoiceDao invoiceDao;
+
+    @BeforeClass(groups = "slow")
+    public void beforeClass() throws Exception {
+        super.beforeClass();
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply beatrix/src/test/java/org/killbill/billing/beatrix/integration/usage/TestWithoutZeroUsageItems.java:60]: We should talk, I am not sure I really understand this.

File: invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceItemGenerator.java [hunk 4]
```diff
@@ -37,6 +38,10 @@
 import org.killbill.billing.junction.BillingEventSet;
 import org.slf4j.Logger;
 
+import com.google.common.base.Predicate;
```
[the remark this answers is not in the document]
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/generator/InvoiceItemGenerator.java:41]: ack, will remove.

File: invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java (hunk 2 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java:656]: There is no good solution. My 2021 view is, that if we come from api, we should throw, but if we come from bus event, we should really park the account as the state becomes inconsistent (but this is not what we do prior my change I think, so as far as I can tell, I tied to current behavior). Let's talk?

**Review by pierre — APPROVED** [review]

**Pull request merged**